### PR TITLE
feat: use RigidityEnv for Rigidity

### PR DESCRIPTION
**Pull request by mlutze**
related to #3715

### CODE REVIEW

**Comment by magnus-madsen** [comment]
What about this PR?

**Comment by mlutze** [comment]
This is the followup to the other one. It contains changes that have a higher risk of breaking things, so I've separated it.

**Comment by mlutze** [comment]
I wanted to be sure we would not see regression in error messages with this change, so I ran some tests.

| VarText type | master count | feature count |
| --- | --- | --- |
| Source | 83 | 85 |
| Fallback | 194 | 196 |
| Absent | 131 | 127 |

It seems things actually improve in that area as well.

**Comment by magnus-madsen** [comment]
> I wanted to be sure we would not see regression in error messages with this change, so I ran some tests.
> VarText type 	master count 	feature count
> Source 	83 	85
> Fallback 	194 	196
> Absent 	131 	127
> 
> It seems things actually improve in that area as well.

Do you think we can get rid of the VarText now?

Is this the PR that reduces the number of ?syms in with --doc to zero?

**Comment by magnus-madsen** [comment]
I think there was a hack in monomorph to make all variables flexible. Is that already gone?

**Comment by magnus-madsen** [comment]
What about removing rigidity from type variables?

**Comment by mlutze** [comment]
> Do you think we can get rid of the VarText now?

Unfortunately this really doesn't change anything for error messages, so VarText situation is the same as before.

> Is this the PR that reduces the number of ?syms in with --doc to zero?

It is! I will request a !!!jar so you can check for yourself to be sure I am not insane.

> I think there was a hack in monomorph to make all variables flexible. Is that already gone?

I have just removed it.

> What about removing rigidity from type variables?

We cannot do this yet, since `Regions` still depends on this field.

**Comment by mlutze** [comment]
I'm getting some of the same on master. Somewhere we must not be calling the smart constructors.

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala [hunk 1]
```diff
@@ -226,12 +226,10 @@ object Typer {
...
           val initialSubst = getSubstFromParams(fparams0)
+          val initialRenv = getRigidityFromParams(fparams0)
 
-          run(initialSubst, RigidityEnv.empty) match { // MATT renv
-            case Ok((subst0, renv0, (partialTconstrs, partialType))) =>
-
-              // propogate the type variable names
-              val subst = subst0.propagate
+          run(initialSubst, initialRenv) match {
+            case Ok((subst, renv0, (partialTconstrs, partialType))) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Typer.scala:232]: rename subst to subst0?

File: main/src/ca/uwaterloo/flix/language/phase/unification/Substitution.scala [hunk 2]
```diff
@@ -166,94 +163,4 @@ case class Substitution(m: Map[Symbol.TypeVarSym, Type]) {
 
     Substitution(newTypeMap.toMap) ++ this
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/Substitution.scala:164]: Maybe we should rename Substitution to `TypeEnv` (to mirror RigidityEnv) at some point.

File: main/src/ca/uwaterloo/flix/language/phase/unification/Substitution.scala [hunk 3]
```diff
@@ -166,94 +163,4 @@ case class Substitution(m: Map[Symbol.TypeVarSym, Type]) {
...
   }
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/Substitution.scala:165]: Nice to get rid of all this!

**Pull request merged**